server/DataStorageController.js

```diff
@@ -0,0 +1,169 @@
+
+// Default DB parameters
+require('dotenv-safe').config();
+const host = process.env.DB_HOST;
+const user = process.env.DB_USER;
+const password = process.env.DB_PASSWORD;
+const database = process.env.DB_DATABASE;
+
+// Establish a connection to the database
+const mySQL = require('mysql')
+const con = mySQL.createConnection({
+	'host': host,
+	'user': user,
+	'password': password,
+	'database': database
+})
+con.connect(function(err){
+	if (err) throw err
+	con.query('CREATE DATABASE IF NOT EXISTS ' + database, function(err, result){
+		if (err) throw err
+	})
+})
+
+// Import database from file
+const Importer = require('mysql-import');
+const importer = new Importer({host, user, password, database});
+function sqldumpImporter(){
+	importer.import('./' + database + '.sql').then(()=>{
+		console.log('SQL file imported.');
+	}).catch(err=>{
+		console.error(err);
+	});
+}
+
+// Export database to file
+var db_file_writer = require('mysqldump');
+function sqldumpexporter(){
+	db_file_writer({
+		connection: {
+			host: host,
+			user: user,
+			password: password,
+			database: database,
+		},
+		dumpToFile: './' + database + '.sql',
+	});
+}
+
+function getCategories(){
+	con.query("SELECT * FROM categories", function (err, result, fields){
+		if(err)
+		{
+			con.query("DROP TABLE IF EXISTS questions", function (err, result){
+				if (err) throw err;
+			});
+			con.query("DROP TABLE IF EXISTS categories", function (err, result){
+				if (err) throw err;
+			});
+
+			// Empty or bad database
+			sqldumpImporter();
+
+			con.query("SELECT * FROM categories", function (err, result, fields){
+				if (err) throw err;
+				categoryList = result;
+			});
+		}
+		else
+		{
+			categoryList = result;
+		}
+	});
+}
+
+function getQuestions(){
+	con.query("SELECT * FROM questions", function (err, result, fields){
+		if(err)
+		{
+			con.query("DROP TABLE IF EXISTS questions", function (err, result){
+				if (err) throw err;
+			});
+			con.query("DROP TABLE IF EXISTS categories", function (err, result){
+				if (err) throw err;
+			});
+
+			// Empty or bad database
+			sqldumpImporter();
+
+			con.query("SELECT * FROM questions", function (err, result, fields){
+				if (err) throw err;
+				questionList = result;
+			});
+		}
+		else
+		{
+			questionList = result;
+		}
+	});
+}
+
+var questionList;
+getQuestions();
+var categoryList;
+getCategories();
+
+// Exports
+
+module.exports.readAQuestion = function(){
+	return questionList[Math.random() * questionList.length];
+}
+
+module.exports.updateCategoryName = function(categoryID, newCategoryName){
+	con.query("UPDATE categories SET name = '" + newCategoryName + "' WHERE id = " + categoryID + ";", function (err, result){
+		if (err) throw err;
+		sqldumpexporter();
+		getCategories();
+	});
+}
+
+module.exports.updateCategoryColor = function(categoryID, newCategoryColor){
+	con.query("UPDATE categories SET color = '" + newCategoryColor + "' WHERE id = " + categoryID + ";", function (err, result){
```

Comment:
This is fine for this use case because we don't have any sensitive data, but concatenating the string for queries allows for sql injection. Just for future reference https://www.acunetix.com/websitesecurity/sql-injection/